src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -375,52 +376,58 @@ private async Task<Project> UpdateProjectInfoAsync(Project project, Checksum inf
+                    project = project.Solution.WithProjectOutputRefFilePath(projectId, newProjectAttributes.OutputRefFilePath).GetRequiredProject(projectId);
                 }
 
                 if (project.State.ProjectInfo.Attributes.CompilationOutputInfo != newProjectAttributes.CompilationOutputInfo)
                 {
-                    project = project.Solution.WithProjectCompilationOutputInfo(project.Id, newProjectAttributes.CompilationOutputInfo).GetProject(project.Id)!;
+                    project = project.Solution.WithProjectCompilationOutputInfo(project.Id, newProjectAttributes.CompilationOutputInfo).GetRequiredProject(project.Id);
                 }
 
                 if (project.State.ProjectInfo.Attributes.DefaultNamespace != newProjectAttributes.DefaultNamespace)
                 {
-                    project = project.Solution.WithProjectDefaultNamespace(projectId, newProjectAttributes.DefaultNamespace).GetProject(projectId)!;
+                    project = project.Solution.WithProjectDefaultNamespace(projectId, newProjectAttributes.DefaultNamespace).GetRequiredProject(projectId);
                 }
 
                 if (project.State.ProjectInfo.Attributes.HasAllInformation != newProjectAttributes.HasAllInformation)
                 {
-                    project = project.Solution.WithHasAllInformation(projectId, newProjectAttributes.HasAllInformation).GetProject(projectId)!;
+                    project = project.Solution.WithHasAllInformation(projectId, newProjectAttributes.HasAllInformation).GetRequiredProject(projectId);
                 }
 
                 if (project.State.ProjectInfo.Attributes.RunAnalyzers != newProjectAttributes.RunAnalyzers)
                 {
-                    project = project.Solution.WithRunAnalyzers(projectId, newProjectAttributes.RunAnalyzers).GetProject(projectId)!;
+                    project = project.Solution.WithRunAnalyzers(projectId, newProjectAttributes.RunAnalyzers).GetRequiredProject(projectId);
                 }
 
                 if (project.State.ProjectInfo.Attributes.ChecksumAlgorithm != newProjectAttributes.ChecksumAlgorithm)
                 {
-                    project = project.Solution.WithProjectChecksumAlgorithm(projectId, newProjectAttributes.ChecksumAlgorithm).GetProject(projectId)!;
+                    project = project.Solution.WithProjectChecksumAlgorithm(projectId, newProjectAttributes.ChecksumAlgorithm).GetRequiredProject(projectId);
+                }
+
+                if (project.State.ProjectInfo.Attributes.SourceGeneratorExecutionVersion != newProjectAttributes.SourceGeneratorExecutionVersion)
+                {
+                    project = project.Solution.WithSourceGeneratorVersions(
+                        FrozenDictionary.ToFrozenDictionary([new KeyValuePair<ProjectId, SourceGeneratorExecutionVersion>(projectId, newProjectAttributes.SourceGeneratorExecutionVersion)]), cancellationToken).GetRequiredProject(projectId);
```

Comment:
this is where we sync the SG execution version from VS to oop.

Reply:
@jasonmalinowski i'm accomplishign all of this by piggybacking this SG-execution-version info within the project-state itself.  That way it automatically syncs over, and we can trivially handle updating the remote side (and its compilation trackers) when it changes.

however, fundamentally, having this info on ProjectState is a bit wonky, and it's unrelated to the actual content/options for projects/documents.  A cleaner place for this to live, imo, would be as an additional map at the SolutionCompilationState level.  So that SolutionCopilationState would have 3 maps:

1. ProjectId->ProjectState
2. ProjectId->CompilationTracker
3. ProjectId->SGExecutionVersion

(or, alternatively, we could put the SGExecutionVersion in the Compilation tracker).

However, if we did that, we'd then need to ensure that when syncing the SolutionCompilationState over to OOP that we could checksum and pass this information along as part of that *high level* sync operation.  We'd also have to make sure that we properly kept that map in sync (for example, when adding/removing projects).

Keeping this info at the project-state level just made a lot of this simpler, even if it's not precisely the most appropriate place for it.

Reply:
The above can be ignored.  I moved to the model where this is all at teh solution compilation level now, not the solution/project-state level.